### ISSUE TITLE
Low performance if using Nginx as a reverse proxy

### ISSUE THREAD

**Issue by vandergrafgenerator** [walkor/Workerman#691]
Workerman http i get 48k requests per second.

Nginx as a reverse proxy for Workerman i only get 31k requests per second.

Am i doing anything wrong?

**nginx.conf**

`user root;
error_log stderr error;
worker_processes auto;

events {
  use epoll;
  worker_connections 16384;
  multi_accept on;
}

http {
  access_log       off;
  sendfile         on;
  tcp_nopush       on;
  tcp_nodelay      on;
  etag off;
  keepalive_requests 10000000;
  upstream app {
    server unix:/dev/shm/app.sock max_fails=0;
  }
  server {
    listen 0.0.0.0:8080 default_server;
    location / {
      proxy_pass http://app;
      proxy_ignore_client_abort on;
    }
  }
}`

**app.php**

`
<?php

use Workerman\Worker;

require_once __DIR__ . '/vendor/autoload.php';

$unix = '/dev/shm/app.sock';
$http_worker = new Worker("unix://$unix");
$http_worker->protocol = "Workerman\\Protocols\\Http"; 

#############################################
#############################################

$http_worker->count = 6;

$http_worker->onMessage = function ($connection, $request) {
 
    $connection->send("Hello World");
};

Worker::runAll();
`

**Comment by walkor** [comment]
Nginx consumes some CPU, because the CPU resources allocated to workerman become less, the QPS will be reduced. However, for most services the performance is still excessive, and the loss of performance by nginx is negligible compared with the loss of business logic.

**Comment by vandergrafgenerator** [comment]
@walkor I believe you are correct. I just verified that that according to htop Workerman cpu usage is about _~30% of each cpu core._ But when running Nginx as a reverse proxy to Workerman cpu usage jumps to about _~90% of each cpu core._

 I wanted to use a reverse proxy Nginx because of attacks like _slow client attack_ and other kinds of ddos. And also to leave static files handling to Nginx.

Do you have any suggestion of an alternative to Nginx for better performance?

**Comment by joanhey** [comment]
Try with `proxy_buffering off;`

```
location / {
      proxy_pass http://app;
      proxy_buffering off;
      proxy_ignore_client_abort on;
    }
```

Without buffers, data is sent from the proxied server and immediately begins to be transmitted to the client. If the clients are assumed to be fast, buffering can be turned off in order to get the data to the client as soon as possible. With buffers, the Nginx proxy will temporarily store the backend’s response and then feed this data to the client. If the client is slow, this allows the Nginx server to close the connection to the backend sooner. It can then handle distributing the data to the client at whatever pace is possible.

Also nginx can terminate HTTPS traffic from clients, relieving your upstream web and application servers of the computational load of SSL/TLS.

And you can join different apps in the same domain.
```
server {
    listen       ...;
    ...
    location / {
        proxy_pass http://127.0.0.1:8080;
    }
    
    location /blog {
        rewrite ^/blog(.*) /$1 break;
        proxy_pass http://127.0.0.1:8181;
    }

    location /mail {
        rewrite ^/mail(.*) /$1 break;
        proxy_pass http://127.0.0.1:8282;
    }
    ...
}
```

**Comment by joanhey** [comment]
Nigns is really fast.
But it's normal to have a penalty for the extra job.

The only other proxy server that I could recommend is Apache Traffic Server.
https://trafficserver.apache.org/
But I don't know of any benchmark about both.

**Comment by joanhey** [comment]
I found a good benchmark and info about proxy servers.
https://static.usenix.org/events/lisa11/tech/slides/hedstrom.pdf

Apparently, Apache Traffic Server is faster. But not necessarily used in the same server.
And remember that the configuration is more difficult.

**Comment by joanhey** [comment]
Try to add keepalive to the upstream, to gain more performance and less .
We have to explicitly enable this setting in Nginx so it does keepalive connections to the upstream it’s connecting to.

```
upstream app {
  # The keepalive parameter sets the maximum number of idle keepalive connections
  # to upstream servers that are preserved in the cache of each worker process. When
  # this number is exceeded, the least recently used connections are closed.
  keepalive 100;

  server unix:/dev/shm/app.sock max_fails=0;
  }
```

You are using unix sockets so the benefit will be less.

Read this article: https://ma.ttias.be/enable-keepalive-connections-in-nginx-upstream-proxy-configurations/

**Comment by joanhey** [comment]
About your nginx config:

It's better to use `multi_accept off` (default) and don't force `epoll `, because nginx will by default use the most efficient method.

**Comment by joanhey** [comment]
You are testing a **hello world!**
If you send a large output, will be better to have `gzip` or `brotli` compression delegated to nginx.

**Comment by vandergrafgenerator** [comment]
@joanhey Thanks for all the information. For now i only tested _proxy_buffering off_ , it increased performance by ~2% only. I believe Nginx may be redirecting to SSL as you said? The CPU usage is too high. I am unfamiliar with Nginx, i will try all your suggestions later and update the results i get here.

**Comment by vandergrafgenerator** [comment]
@joanhey I did all you recommended except that i did not add `proxy_buffering off;` because it seemed to me it could facilitate slow client attack? Is that assumption correct?

It got a huge performance improvement: It went from 31k to 40k requests per second which is close to the 48k i get with pure Workerman. There was algo a huge improvement in CPU usage, it went down from 90% to 36% which also close to the 30% i get from pure Workerman. I am benchmarking with 100 concurrent users.

With so much idle CPU i suppose there may be room to improve even more?  

That is how the nginx.conf ended:

`user root;

worker_cpu_affinity auto;
worker_processes auto;
timer_resolution 1s;

events {
 
  worker_connections 32768;
  multi_accept off;
}

http {
  access_log       off;
  sendfile         off;
  tcp_nopush       off;
  tcp_nodelay      on;
  etag off;
  server_tokens off;
  msie_padding off;
  keepalive_disable none; 
  keepalive_requests 1000;
  upstream app {
    keepalive 100;
    server unix:/dev/shm/app.sock max_fails=0;
  }
  server {
    listen 0.0.0.0:8080 default_server reuseport;
    location / {
      proxy_pass http://app;
      proxy_ignore_client_abort on;
    }
  }
}`

**Comment by xpader** [comment]
I think may the key point is keepalive?

**Comment by joanhey** [comment]
The keepalive in upstream is very important for a TCP connection but insignificant using Unix sockets.
The big problem was the high number of keealive_requests.

**Comment by joanhey** [comment]
@vandergrafgenerator 

> @joanhey I did all you recommended except that i did not add proxy_buffering off; because it seemed to me it could facilitate slow client attack? Is that assumption correct?

It's correct

Please don't use `user root:` use `user www-data;`

**Comment by vandergrafgenerator** [comment]
> @vandergrafgenerator
> 
> > @joanhey I did all you recommended except that i did not add proxy_buffering off; because it seemed to me it could facilitate slow client attack? Is that assumption correct?
> 
> It's correct
> 
> Please don't use `user root;` use `user www-data;`
> 
> Which program do you use for benchmark ? `ab` or `wrk` or .. ?

I did use Apache Jmeter: 100 users(threads), zero ramp up time, infinite loop.

**Comment by raghuveer** [comment]
@walkor do you suggest Nginx as reverse proxy on a different machine to do SSL offloading while Workerman does it's job as http server then?

Do you think performance be the same then? with iptables restricting access to workerman based http server from nginx reverse proxy ip address only kind of?

please share your thoughts in performance perspective

thank you

**Comment by walkor** [comment]
Yes, if SSL is necessary, using nginx reverse proxy SSL is a very good idea in my opinion.

In terms of performance, running nginx on a different machine does not affect the performance of workerman. However, as a whole, the throughput does not just depend on the workerman. The specific throughput needs to be tested.

**Comment by raghuveer** [comment]
fully agree, to test all scenarios and document this sometime over Workerman v5.

**Comment by joanhey** [comment]
Sorry, I never remember to add that to proxy connections.

Add to the location:
```
proxy_http_version 1.1;
proxy_set_header Connection "";
```

By default in nginx, the proxy upstream use `http 1.0` (so no real keepalive).
And change the `Connection` header to not listen the `close`, to maintain open the upstream connection (be carefull in some apps).

@vandergrafgenerator Could you benhmark these changes and inform us. 
Thank you.

**Comment by ewwink** [comment]
sorry to bump, but I found better option for reverse proxy, I use [Caddy](https://github.com/caddyserver/caddy) and below is the benchmark result, it running on Windows and 1 Workerman process


Direct/Without Proxy

```yaml
command: oha -n 5000 -c 50 http://target:8080/
Summary:
  Success rate: 100.00%
  Total:        113.3569 ms
  Slowest:      30.1120 ms
  Fastest:      0.0997 ms
  Average:      1.1103 ms
  Requests/sec: 44108
```

Nginx Proxy Pass with worker_connections 1024 (higher has no effect on Windows)

```yaml
command: oha -n 5000 -c 50 http://target/
Summary:
  Success rate: 100.00%
  Total:        1795.6794 ms
  Slowest:      29.9673 ms
  Fastest:      9.2154 ms
  Average:      17.9009 ms
  Requests/sec: 2784
```

Caddy Proxy Pass (10x faster than Nginx)

```yaml
command: oha -n 5000 -c 50 http://target/
Summary:
  Success rate: 100.00%
  Total:        195.7550 ms
  Slowest:      52.2721 ms
  Fastest:      0.3297 ms
  Average:      1.9288 ms
  Requests/sec: 25542
```
